File: ruby/rbsearch/searchresult.rb
# frozen_string_literal: true

# SearchResult - encapsulates a search result
class SearchResult
  attr_accessor :pattern
  attr_accessor :file
  attr_accessor :linenum
  attr_accessor :match_start_index
  attr_accessor :match_end_index
  attr_accessor :line
  attr_accessor :lines_before
  attr_accessor :lines_after

  SEPARATOR_LEN = 80

  def initialize(pattern, file, linenum=0, match_start_index=0,
                 match_end_index=0, line='', lines_before=[], lines_after=[])
    @pattern = pattern
    @file = file
    @linenum = linenum
    @match_start_index = match_start_index
    @match_end_index = match_end_index
    @line = line
    @lines_before = lines_before
    @lines_after = lines_after
    @maxlinelength = 150
  end

  def to_s
    if @lines_before.empty? && @lines_after.empty?
      singleline_s
    else
      multiline_s
    end
  end

  private

  def singleline_s
    s = @file.to_s
    if @linenum.positive? && !@line.empty?
      s << ": #{@linenum}: [#{@match_start_index}:#{@match_end_index}]:"
      s << " #{@line.strip}"
    else
      s << " matches at [#{@match_start_index}:#{@match_end_index}]"
    end
    s
  end

  def format_matching_line
    formatted = @line
    linelength = @line.length
    matchlength = @match_end_index - @match_start_index
    if linelength > @maxlinelength
      adjusted_maxlength = @maxlinelength - matchlength
      before_index = @match_start_index
      if @match_start_index > 0
        before_index = before_index - (adjusted_maxlength / 4)
        if before_index < 0
          before_index = 0
        end
      end
      adjusted_maxlength = adjusted_maxlength - (@match_start_index - before_index)
      after_index = @match_end_index + adjusted_maxlength
      if after_index > linelength
        after_index = linelength
      end
      before = ''
      if before_index > 3
        before = '...'
        before_index += 3
      end
      after = ''
      if after_index < linelength - 3
        after = '...'
        after_index -= 3
      end
      formatted = before + @line[before_index..after_index] + after
    end
    formatted.strip
  end

  def linenum_padding
    max_linenum = @linenum + @lines_after.length
    max_linenum.to_s.length
  end

  def strip_newlines(s)
    s = s[0..-2] while s.end_with? "\n", "\r"
    s
  end

  def multiline_s
    s = '=' * SEPARATOR_LEN + "\n"
    s << "#{@file}: #{@linenum}: [#{@match_start_index}:#{@match_end_index}]\n"
    s << '-' * SEPARATOR_LEN + "\n"
    line_format = " %%%dd | %%s\n" % [linenum_padding]
    current_linenum = @linenum
    unless @lines_before.empty?
      current_linenum -= @lines_before.length
      @lines_before.each do |line_before|
        s << ' ' + line_format % [current_linenum, strip_newlines(line_before)]
        current_linenum += 1
      end
    end
    s << '>' + line_format % [current_linenum, strip_newlines(@line)]
    unless @lines_after.empty?
      current_linenum += 1
      @lines_after.each do |line_after|
        s << ' ' + line_format % [current_linenum, strip_newlines(line_after)]
        current_linenum += 1
      end
    end
    s + "\n"
  end
end

File: node/jssearch/src/searchresult.js
/*
 * searchresults.js
 *
 * SearchResult class represents a search result
 */

function SearchResult(pattern, filename, linenum, matchStartIndex, matchEndIndex,
    line, linesBefore, linesAfter) {
    var self = this;
    self.pattern = pattern;
    self.filename = filename;
    self.linenum = linenum;
    self.matchStartIndex = matchStartIndex;
    self.matchEndIndex = matchEndIndex;
    self.line = line;
    self.linesBefore = linesBefore;
    self.linesAfter = linesAfter;
    self.maxLineLength = 150;

    self.toString = function () {
        return self.linesBefore.length + self.linesAfter.length > 0 ?
            multiLineToString() : singleLineToString();
    };

    var singleLineToString = function () {
        var s = self.filename;
        if (self.linenum && self.line) {
            s += ': ' + self.linenum + ': [' + self.matchStartIndex + ':' +
                self.matchEndIndex +']: ' + formatMatchingLine();
        } else {
            s += ' matches at [' + self.matchStartIndex + ':' +
                self.matchEndIndex + ']';
        }
        return s;
    };

    var lineNumPadding = function() {
        var maxLineNum = self.linenum + self.linesAfter.length;
        return ("" + maxLineNum).length;
    };

    var padLeft = function(s, i) {
        p = "" + s;
        while (p.length < i) {
            p = " " + p;
        }
        return p;
    };

    var trimRight = function(s) {
        var trimmed = s.replace(/[\r\n]+$/, '');
        return trimmed;
    };

    var multiLineToString = function () {
        var s = Array(81).join("=") + "\n";
        s += "{0}: {1}: ".format(self.filename, self.linenum);
        s += "[{0}:{1}]\n".format(self.matchStartIndex, self.matchEndIndex);
        s += Array(81).join("-") + "\n";
        var currentLineNum = self.linenum;
        var numPadding = lineNumPadding();
        if (self.linesBefore.length > 0) {
            currentLineNum = currentLineNum - self.linesBefore.length;
            for (var i = 0; i < self.linesBefore.length; i++) {
                var b = trimRight(self.linesBefore[i]);
                s += "  " + padLeft(currentLineNum, numPadding) + " | " + b + "\n";
                currentLineNum++;
            }
        }
        s += "> " + padLeft(currentLineNum, numPadding) + " | " +
            trimRight(self.line) + "\n";
        if (self.linesAfter.length > 0) {
            currentLineNum++;
            for (var i = 0; i < self.linesAfter.length; i++) {
                var a = trimRight(self.linesAfter[i]);
                s += "  " + padLeft(currentLineNum, numPadding) + " | " + a + "\n";
                currentLineNum++;
            }
        }
        return s;
    };

    var formatMatchingLine = function () {
        var formatted = self.line;
        var lineLength = self.line.length;
        var matchLength = self.matchEndIndex - self.matchStartIndex;
        if (lineLength > self.maxLineLength) {
            var adjustedMaxLength = self.maxLineLength - matchLength;
            var beforeIndex = self.matchStartIndex;
            if (self.matchStartIndex > 0) {
                beforeIndex = beforeIndex - (adjustedMaxLength / 4);
                if (beforeIndex < 0)
                    beforeIndex = 0;
            }
            adjustedMaxLength = adjustedMaxLength - (self.matchStartIndex - beforeIndex);
            var afterIndex = self.matchEndIndex + adjustedMaxLength;
            if (afterIndex > lineLength)
                afterIndex = lineLength;

            var before = '';
            if (beforeIndex > 3) {
                before = '...';
                beforeIndex += 3;
            }
            var after = '';
            if (afterIndex < lineLength - 3) {
                after = '...';
                afterIndex -= 3;
            }
            formatted = before + self.line.substring(beforeIndex, afterIndex) + after;
        }
        return formatted.trim();
    };
}

exports.SearchResult = SearchResult;
